File: docs/app/weather.js
// Weather alerts for the Gulf. Shared by the page (imported as a module) and by the
// service worker (importScripts), so it declares no imports or exports and sets one global.
//
// Data: Open-Meteo forecast and air quality APIs (free, no key), sent only an approximate
// location, and only when the person turns weather alerts on.
//
// Dust is judged against the place's own recent history: Kuwait and Riyadh usually carry
// ten times the dust of Muscat, so one fixed number would alert every day in one city and
// never in another. A day is dusty when it beats 90% of the last 60 days there.
(function (root) {
  const FORECAST = 'https://api.open-meteo.com/v1/forecast';
  const AIR = 'https://air-quality-api.open-meteo.com/v1/air-quality';

  const LIMITS = {
    dustFloor: 150, // never call a day dusty below this PM10 daily maximum, in micrograms per cubic metre
    rainMm: 0.5,
    rainChance: 50,
    heat: 48, // degrees C, daily maximum
    cold: 4, // degrees C, daily minimum
    gust: 60, // km/h
  };
  const ORDER = { dust_heavy: 0, rain: 1, wind: 2, dust: 3, heat: 4, cold: 5 };

  function urls(lat, lon) {
    const at = `latitude=${lat}&longitude=${lon}&timezone=auto`;
    return {
      forecast: `${FORECAST}?${at}&daily=temperature_2m_max,temperature_2m_min,precipitation_sum,precipitation_probability_max,wind_gusts_10m_max&forecast_days=3`,
      air: `${AIR}?${at}&hourly=pm10&past_days=60&forecast_days=3`,
    };
  }

  function percentile(sorted, f) {
    if (!sorted.length) return null;
    return sorted[Math.min(sorted.length - 1, Math.floor(f * (sorted.length - 1)))];
  }

  // Both responses into one small summary that is cheap to keep.
  function summarize(forecast, air) {
    const dust = {};
    const hourly = (air && air.hourly) || {};
    (hourly.time || []).forEach((t, i) => {
      const v = hourly.pm10 ? hourly.pm10[i] : null;
      if (v == null) return;
      const day = t.slice(0, 10);
      dust[day] = Math.max(dust[day] || 0, v);
    });
    const f = (forecast && forecast.daily) || {};
    const days = (f.time || []).map((date, i) => ({
      date,
      tmax: f.temperature_2m_max ? f.temperature_2m_max[i] : null,
      tmin: f.temperature_2m_min ? f.temperature_2m_min[i] : null,
      rain: f.precipitation_sum ? f.precipitation_sum[i] : null,
      chance: f.precipitation_probability_max ? f.precipitation_probability_max[i] : null,
      gust: f.wind_gusts_10m_max ? f.wind_gusts_10m_max[i] : null,
      dust: dust[date] != null ? Math.round(dust[date]) : null,
    }));
    const first = days.length ? days[0].date : '9999-12-31';
    const history = Object.keys(dust).filter((d) => d < first).map((d) => dust[d]).sort((a, b) => a - b);
    const p90 = percentile(history, 0.9);
    const p97 = percentile(history, 0.97);
    const dusty = Math.max(LIMITS.dustFloor, p90 == null ? Infinity : p90);
    const heavy = Math.max(dusty * 1.25, p97 == null ? Infinity : p97);
    return { days, dusty: Math.round(dusty), heavy: Math.round(heavy), history: history.length };
  }

  // Alerts for today (day 0) and tomorrow (day 1), most serious first.
  function alerts(summary, today) {
    const out = [];
    if (!summary || !summary.days) return out;
    for (const [dayIndex, date] of [[0, today], [1, nextDay(today)]]) {
      const d = summary.days.find((x) => x.date === date);
      if (!d) continue;
      if (d.dust != null && summary.history >= 20) {
        if (d.dust >= summary.heavy) out.push({ kind: 'dust_heavy', day: dayIndex });
        else if (d.dust >= summary.dusty) out.push({ kind: 'dust', day: dayIndex });
      }
      if ((d.rain != null && d.rain >= LIMITS.rainMm) || (d.chance != null && d.chance >= LIMITS.rainChance)) out.push({ kind: 'rain', day: dayIndex });
      if (d.gust != null && d.gust >= LIMITS.gust) out.push({ kind: 'wind', day: dayIndex });
      if (d.tmax != null && d.tmax >= LIMITS.heat) out.push({ kind: 'heat', day: dayIndex, value: Math.round(d.tmax) });
      if (d.tmin != null && d.tmin <= LIMITS.cold) out.push({ kind: 'cold', day: dayIndex, value: Math.round(d.tmin) });
    }
    return out.sort((a, b) => a.day - b.day || ORDER[a.kind] - ORDER[b.kind]);
  }

  function nextDay(iso) {
    const [y, m, d] = iso.split('-').map(Number);
    const t = new Date(Date.UTC(y, m - 1, d + 1));
    return t.toISOString().slice(0, 10);
  }

  async function fetchWeather(lat, lon, fetchImpl, timeoutMs) {
    const u = urls(lat, lon);
    const ctl = typeof AbortController !== 'undefined' ? new AbortController() : null;
    const timer = ctl ? setTimeout(() => ctl.abort(), timeoutMs || 10000) : null;
    try {
      const opts = ctl ? { signal: ctl.signal, credentials: 'omit', referrerPolicy: 'no-referrer' } : {};
      const [a, b] = await Promise.all([fetchImpl(u.forecast, opts), fetchImpl(u.air, opts)]);
      if (!a.ok || !b.ok) throw new Error('weather ' + a.status + ' ' + b.status);
      return summarize(await a.json(), await b.json());
    } finally {
      if (timer) clearTimeout(timer);
    }
  }

  root.NokhathaWeather = { LIMITS, urls, summarize, alerts, fetchWeather, nextDay };
})(typeof self !== 'undefined' ? self : globalThis);

File: android/app/src/main/java/com/eworldq8/nokhatha/engine/Weather.kt
// Weather alerts, the one thing the app fetches from outside the device, and only when the person
// turns it on: dust, rain, wind, heat and cold for the area they chose, from Open-Meteo. The
// same thresholds as docs/app/weather.js, so the web and the phone warn on the same days.
package com.eworldq8.nokhatha.engine

import org.json.JSONArray
import org.json.JSONObject

object Weather {
    const val FORECAST = "https://api.open-meteo.com/v1/forecast"
    const val AIR = "https://air-quality-api.open-meteo.com/v1/air-quality"
    const val DUST_FLOOR = 150.0 // never call a day dusty below this PM10 daily maximum
    const val RAIN_MM = 0.5
    const val RAIN_CHANCE = 50.0
    const val HEAT = 48.0
    const val COLD = 4.0
    const val GUST = 60.0
    private val order = listOf("dust_heavy", "rain", "wind", "dust", "heat", "cold")

    data class DayWx(val date: String, val tmax: Double?, val tmin: Double?, val rain: Double?, val chance: Double?, val gust: Double?, val dust: Int?)
    data class Summary(val days: List<DayWx>, val dusty: Int, val heavy: Int, val history: Int)
    data class Alert(val kind: String, val day: Int, val value: Int? = null)

    fun urls(lat: Double, lon: Double): Pair<String, String> {
        val at = "latitude=${num(lat)}&longitude=${num(lon)}&timezone=auto"
        return Pair(
            "$FORECAST?$at&daily=temperature_2m_max,temperature_2m_min,precipitation_sum,precipitation_probability_max,wind_gusts_10m_max&forecast_days=3",
            "$AIR?$at&hourly=pm10&past_days=60&forecast_days=3",
        )
    }

    /** Numbers as JavaScript prints them: 29.4 stays 29.4 and 48.0 becomes 48. */
    private fun num(v: Double): String = if (v == Math.floor(v) && !v.isInfinite()) v.toLong().toString() else v.toString()

    private fun percentile(sorted: List<Double>, f: Double): Double? =
        if (sorted.isEmpty()) null else sorted[minOf(sorted.size - 1, Math.floor(f * (sorted.size - 1)).toInt())]

    private fun JSONArray?.num(i: Int): Double? = if (this == null || i >= length() || isNull(i)) null else optDouble(i)

    /** The daily maximum PM10 per day, the forecast days, and the dust thresholds from the last 60 days. */
    fun summarize(forecast: JSONObject?, air: JSONObject?): Summary {
        val dust = LinkedHashMap<String, Double>()
        val hourly = air?.optJSONObject("hourly")
        val times = hourly?.optJSONArray("time")
        val pm10 = hourly?.optJSONArray("pm10")
        if (times != null) for (i in 0 until times.length()) {
            val v = pm10.num(i) ?: continue
            val day = times.getString(i).take(10)
            dust[day] = maxOf(dust[day] ?: 0.0, v)
        }
        val f = forecast?.optJSONObject("daily")
        val dates = f?.optJSONArray("time")
        val days = ArrayList<DayWx>()
        if (dates != null) for (i in 0 until dates.length()) {
            val date = dates.getString(i)
            days += DayWx(date, f.optJSONArray("temperature_2m_max").num(i), f.optJSONArray("temperature_2m_min").num(i),
                f.optJSONArray("precipitation_sum").num(i), f.optJSONArray("precipitation_probability_max").num(i),
                f.optJSONArray("wind_gusts_10m_max").num(i), dust[date]?.let { Math.round(it).toInt() })
        }
        val first = days.firstOrNull()?.date ?: "9999-12-31"
        val history = dust.filterKeys { it < first }.values.sorted()
        val p90 = percentile(history, 0.9)
        val p97 = percentile(history, 0.97)
        val dusty = maxOf(DUST_FLOOR, p90 ?: Double.POSITIVE_INFINITY)
        val heavy = maxOf(dusty * 1.25, p97 ?: Double.POSITIVE_INFINITY)
        return Summary(days, roundOrMax(dusty), roundOrMax(heavy), history.size)
    }

    private fun roundOrMax(v: Double) = if (v.isInfinite()) Int.MAX_VALUE else Math.round(v).toInt()

    /** Alerts for today and tomorrow, most serious first. */
    fun alerts(s: Summary?, today: Day): List<Alert> {
        if (s == null) return emptyList()
        val out = ArrayList<Alert>()
        for ((dayIndex, date) in listOf(0 to today.iso, 1 to today.plus(1).iso)) {
            val d = s.days.firstOrNull { it.date == date } ?: continue
            if (d.dust != null && s.history >= 20) {
                if (d.dust >= s.heavy) out += Alert("dust_heavy", dayIndex)
                else if (d.dust >= s.dusty) out += Alert("dust", dayIndex)
            }
            if ((d.rain != null && d.rain >= RAIN_MM) || (d.chance != null && d.chance >= RAIN_CHANCE)) out += Alert("rain", dayIndex)
            if (d.gust != null && d.gust >= GUST) out += Alert("wind", dayIndex)
            if (d.tmax != null && d.tmax >= HEAT) out += Alert("heat", dayIndex, Math.round(d.tmax).toInt())
            if (d.tmin != null && d.tmin <= COLD) out += Alert("cold", dayIndex, Math.round(d.tmin).toInt())
        }
        return out.sortedWith(compareBy<Alert> { it.day }.thenBy { order.indexOf(it.kind) })
    }

    // The summary is cached on the device as JSON, so alerts can be shown without the network.

    fun toJson(s: Summary): JSONObject = JSONObject().put("dusty", s.dusty).put("heavy", s.heavy).put("history", s.history)
        .put("days", JSONArray(s.days.map { d ->
            JSONObject().put("date", d.date).put("tmax", d.tmax ?: JSONObject.NULL).put("tmin", d.tmin ?: JSONObject.NULL).put("rain", d.rain ?: JSONObject.NULL)
                .put("chance", d.chance ?: JSONObject.NULL).put("gust", d.gust ?: JSONObject.NULL).put("dust", d.dust ?: JSONObject.NULL)
        }))

    fun fromJson(o: JSONObject): Summary {
        fun JSONObject.d(k: String): Double? = if (has(k) && !isNull(k)) getDouble(k) else null
        val days = o.optJSONArray("days") ?: JSONArray()
        return Summary((0 until days.length()).map { i ->
            val d = days.getJSONObject(i)
            DayWx(d.getString("date"), d.d("tmax"), d.d("tmin"), d.d("rain"), d.d("chance"), d.d("gust"), d.d("dust")?.toInt())
        }, o.getInt("dusty"), o.getInt("heavy"), o.getInt("history"))
    }

    val icons = mapOf("dust" to "dust", "dust_heavy" to "dust", "rain" to "rain", "wind" to "air", "heat" to "thermo", "cold" to "snow")
}

/** The weather setting, part of settings.weather in the shared JSON. */
data class WeatherSetting(val on: Boolean = false, val place: String? = null, val lat: Double? = null, val lon: Double? = null) {
    fun toJson(): JSONObject = JSONObject().put("on", on).apply {
        if (place != null) put("place", place)
        if (lat != null) put("lat", lat)
        if (lon != null) put("lon", lon)
    }

    companion object {
        fun fromJson(o: JSONObject?): WeatherSetting? = o?.let {
            WeatherSetting(it.optBoolean("on", false), if (it.has("place") && !it.isNull("place")) it.getString("place") else null,
                if (it.has("lat") && !it.isNull("lat")) it.getDouble("lat") else null, if (it.has("lon") && !it.isNull("lon")) it.getDouble("lon") else null)
        }
    }
}
